v3/events.go

```diff
@@ -77,6 +78,23 @@ func PollEvents() <-chan Event {
 	return ch
 }
 
+// PollEventsWithContext same as PollEvents() prevents goroutine leaking using context.
+func PollEventsWithContext(ctx context.Context) <-chan Event {
+	ch := make(chan Event)
+	go func() {
+		defer close(ch)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				ch <- convertTermboxEvent(tb.PollEvent())
```

Comment:
```suggestion
			case ch <- convertTermboxEvent(tb.PollEvent()):
```
is better, otherwise while you are waiting for `ch` input, cancelling the context is not efective.

Reply:
@maxatome The core problem is that `tb.PollEvent()` is blocking.
This suggestion would not fix that what you claim as it would just handle a cancel during the write on `ch` *after* an event has been blockingly read.

Reply:
> @maxatome The core problem is that `tb.PollEvent()` is blocking.
> This suggestion would not fix that what you claim as it would just handle a cancel during the write on `ch` _after_ an event has been blockingly read.

Yes, I had this in mind.

Reply:
Do you have any suggestions how to make it better?

Reply:
Good point @dolmen :)

What about a specific goroutine to handle this blocking function?

Something like that (not tested in termui context):
```go
func PollEventsWithContext(ctx context.Context) <-chan Event {
	// tb.PollEvent() is a blocking function, so dedicate a goroutine for it
	tbEvent := make(chan Event)
	go func() {
		defer close(tbEvent)
		for {
			ev := tb.PollEvent()
			if ev.Type == tb.EventInterrupt {
				return
			}
			tbEvent <- convertTermboxEvent(ev)
		}
	}()

	ch := make(chan Event)
	go func() {
		defer close(ch)
		ctxDone := ctx.Done()
		for {
			select {
			case ctxDone:
				// interrupt tb.PollEvent(), this call blocks until the
				// PollEvent function has successfully been interrupted
				tb.Interrupt()
				ctxDone = nil
			case ev := <-tbEvent:
				// Successfully interrupted
				if ev.Type == 0 {
					return
				}
				ch <- ev
			}
		}
	}()
	return ch
}
```

Reply:
Should works, however it looks like over usage of go-routines to me. I would like to suggest a simpler way to do it:
```
func PollEventsWithCancel() (<-chan Event, func()) {
	ch := make(chan Event)
	go func() {
		defer close(ch)
		for {
			ev := tb.PollEvent()
			if ev.Type == tb.EventInterrupt {
				return
			}
			ch <- convertTermboxEvent(ev)
		}
	}()
	return ch, func(){tb.Interrupt()}
}
```

Reply:
Another concern is that the method should be a singleton. In other way behavior is unpredictable and doesn't really make sense.

Reply:
Let me back to my pet proj after a while and see what can be done, any suggestions are welcome.